### PR TITLE
Test Conway CDDL using Cuddle

### PR DESCRIPTION
**Pull request by nc6**
# Description

This PR introduces [cuddle](https://github.com/input-output-hk/cuddle), a Haskell library for declaring and validating CDDL documents.

This is then used to reflect the Conway CDDL specification (in eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs) using a DSL known as Huddle.

libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/Cuddle.hs introduces utilities to run CDDL conformance tests using cuddle. These are currently based on a Huddle declaration, but it would be equally possible to use a native CDDL specification combined with cuddle's CBOR generators.

eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs then introduces a new set of tests based in the Huddle conway specification. The existing test suite is left untouched for now.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs [hunk 1]
```diff
@@ -15,28 +16,59 @@ import Test.Cardano.Ledger.Binary.Cddl (
...
+  huddleRoundTripCborSpec @(TxOut Conway) v "transaction_output"
+  huddleRoundTripAnnCborSpec @(TxBody Conway) v "transaction_body"
+  huddleRoundTripAnnCborSpec @(TxAuxData Conway) v "auxiliary_data"
+  huddleRoundTripAnnCborSpec @(Timelock Conway) v "native_script"
+  huddleRoundTripAnnCborSpec @(Data Conway) v "plutus_data"
+  huddleRoundTripAnnCborSpec @(Script Conway) v "script"
+  huddleRoundTripAnnCborSpec @(TxWits Conway) v "transaction_witness_set"
+  huddleRoundTripAnnCborSpec @(Redeemers Conway) v "redeemers"
+  huddleRoundTripAnnCborSpec @(Tx Conway) v "transaction"
+  pure ()
```
Comment (teodanciu) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs]: Seems unnecessary: 
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 2]
```diff
@@ -0,0 +1,971 @@
...
+      , opt $ idx 4 ==> nonempty_set plutus_data
+      , opt $ idx 5 ==> redeemers
+      , opt $ idx 6 ==> nonempty_set plutus_v2_script
+      , opt $ idx 7 ==> nonempty_set plutus_v3_script
+      ]
+
+-- The real type of  plutus_v1_script, plutus_v2_script and plutus_v3_script is bytes.
+-- However, because we enforce uniqueness when many scripts are supplied,
+-- we need to hack around for tests in order to avoid generating duplicates,
+-- since the cddl tool we use for roundtrip testing doesn't generate distinct collections.
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Is this comment still applicable? it seems like `distinct` has solved the problem that's being described in, in which case we can get rid of it.
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 3]
```diff
@@ -0,0 +1,971 @@
...
+plutus_v1_script :: Rule
+plutus_v1_script =
+  comment
+    ( "The real type of  plutus_v1_script, plutus_v2_script and plutus_v3_script is bytes.\n"
+        <> "However, because we enforce uniqueness when many scripts are supplied,\n"
+        <> "we need to hack around for tests in order to avoid generating duplicates,\n"
+        <> "since the cddl tool we use for roundtrip testing doesn't generate distinct collections.\n"
+    )
+    $ "plutus_v1_script" =:= distinct VBytes
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]:
```suggestion
   "plutus_v 1_script" =:= distinct VBytes
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 4]
```diff
@@ -0,0 +1,971 @@
...
+import Codec.CBOR.Cuddle.Huddle
+import Data.Function (($))
+import Data.Semigroup ((<>))
+import Data.Text qualified as T
+import Data.Word (Word64)
+import GHC.Float (Double)
+import GHC.Num (Integer)
+import GHC.Show (Show (show))
+
+default (Integer, Double)
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Isn't this the default default?  I think we could get rid of this :thinking: 
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 5]
```diff
@@ -0,0 +1,971 @@
...
+
+conway :: Huddle
+conway = collectFrom [block, transaction]
+
+block :: Rule
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Is the plan for the future to define these in Shelley and to reuse them here (where appropriate)?   And like this to replace our cddl-tool that generates examples?

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 6]
```diff
@@ -0,0 +1,971 @@
...
+header_body =
+  "header_body"
+    =:= arr
+      [ "block_number" ==> VUInt
+      , "slot" ==> VUInt
+      , "prev_hash" ==> hash32
+      , "issuer_vkey" ==> vkey
+      , "vrf_vkey" ==> vkey
+      , "vrf_result" ==> vrf_cert
+      , "block_body_size" ==> VUInt
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Are we using `VUInt` to represent all sizes?  Is this the same VUint defined like this in cuddle?
```
data Value
  = VUInt Word64
  | VNInt Word64
   ..
```

How does it work then,  can we define precise sizes for the unsigned integers as we do in the cddl file?

File: libs/cardano-ledger-binary/cardano-ledger-binary.cabal
Comment (teodanciu) [inline libs/cardano-ledger-binary/cardano-ledger-binary.cabal:1]: Do you mind updating the package version  to 1.3.4.0 ?

File: eras/conway/impl/cardano-ledger-conway.cabal [hunk 7]
```diff
@@ -96,6 +97,7 @@ library
         cardano-slotting,
         cardano-strict-containers,
         containers,
+        cuddle,
```
Comment (teodanciu) [inline eras/conway/impl/cardano-ledger-conway.cabal]: `cardano-ledger-binary` version above needs to be updated to >= 1.3.4  (if  [this suggestion is followed](https://github.com/IntersectMBO/cardano-ledger/pull/4437/files#r1662738083))

Comment (lehins) [reply eras/conway/impl/cardano-ledger-conway.cabal]: We don't want libraries to depend on `cuddle`, because neither that library not the CDDL spec that is generated from it is not necessary for cardano-node execution. We only need `cuddle` for testing and producing CDDL spec, so everything `cuddle` related that is reusable should go into `testlib`.

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs [hunk 8]
```diff
@@ -15,28 +16,59 @@ import Test.Cardano.Ledger.Binary.Cddl (
...
+        cddlRoundTripCborSpec @(Datum Conway) v "datum_option"
+        cddlRoundTripAnnCborSpec @(TxWits Conway) v "transaction_witness_set"
+        cddlRoundTripCborSpec @(PParamsUpdate Conway) v "protocol_param_update"
+        cddlRoundTripCborSpec @CostModels v "costmdls"
+        cddlRoundTripAnnCborSpec @(Redeemers Conway) v "redeemers"
+        cddlRoundTripAnnCborSpec @(Tx Conway) v "transaction"
+        cddlRoundTripCborSpec @(VotingProcedure Conway) v "voting_procedure"
+        cddlRoundTripCborSpec @(ProposalProcedure Conway) v "proposal_procedure"
+        cddlRoundTripCborSpec @(GovAction Conway) v "gov_action"
+        cddlRoundTripCborSpec @(TxCert Conway) v "certificate"
```
Comment (lehins) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs]:
```haskell
foo $ 
  do
    bar
```

This is the style we try to avoid in ledger :wink: Also, this diff is totally unnecessary, so let's minimize it by reverting the change.
```suggestion
  describe "CDDL" $ beforeAllCddlFile 3 readConwayCddlFiles $ do
    let v = eraProtVerHigh @Conway
    cddlRoundTripCborSpec @(Value Conway) v "positive_coin"
    cddlRoundTripCborSpec @(Value Conway) v "value"
    cddlRoundTripAnnCborSpec @(TxBody Conway) v "transaction_body"
    cddlRoundTripAnnCborSpec @(TxAuxData Conway) v "auxiliary_data"
    cddlRoundTripAnnCborSpec @(Timelock Conway) v "native_script"
    cddlRoundTripAnnCborSpec @(Data Conway) v "plutus_data"
    cddlRoundTripCborSpec @(TxOut Conway) v "transaction_output"
    cddlRoundTripAnnCborSpec @(Script Conway) v "script"
    cddlRoundTripCborSpec @(Datum Conway) v "datum_option"
    cddlRoundTripAnnCborSpec @(TxWits Conway) v "transaction_witness_set"
    cddlRoundTripCborSpec @(PParamsUpdate Conway) v "protocol_param_update"
    cddlRoundTripCborSpec @CostModels v "costmdls"
    cddlRoundTripAnnCborSpec @(Redeemers Conway) v "redeemers"
    cddlRoundTripAnnCborSpec @(Tx Conway) v "transaction"
    cddlRoundTripCborSpec @(VotingProcedure Conway) v "voting_procedure"
    cddlRoundTripCborSpec @(ProposalProcedure Conway) v "proposal_procedure"
    cddlRoundTripCborSpec @(GovAction Conway) v "gov_action"
    cddlRoundTripCborSpec @(TxCert Conway) v "certificate"
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 9]
```diff
@@ -0,0 +1,971 @@
...
+reg_drep_cert = "reg_drep_cert" =:~ grp [16, a drep_credential, a coin, anchor // VNil]
+
+unreg_drep_cert :: Named Group
+unreg_drep_cert = "unreg_drep_cert" =:~ grp [17, a drep_credential, a coin]
+
+update_drep_cert :: Named Group
+update_drep_cert = "update_drep_cert" =:~ grp [18, a drep_credential, anchor // VNil]
+
+delta_coin :: Rule
+delta_coin = "delta_coin" =:= VUInt
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: `delta_coin` is actually signed
```haskell
delta_coin = "delta_coin" =:= VInt
```

But, it is unused so we can get rid of it. It is unnecessary in the original spec either, so if you can remove it there too that would be helpful.
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs (hunk 6 above)
Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: This is a very important question. We do need ability to specify sizes, eg. [`block_body_size  : uint .size 4`](https://github.com/IntersectMBO/cardano-ledger/blob/f4119c9cd27e69943dcf91d22d0f9905be32fa75/eras/conway/impl/cddl-files/conway.cddl#L34)

I assume it can be achieved with something like this, right?
```suggestion
      , "block_body_size" ==> VUInt `sized` (4 :: Word)
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 10]
```diff
@@ -0,0 +1,971 @@
...
+header :: Rule
+header = "header" =:= arr [a header_body, "body_signature" ==> kes_signature]
+
+header_body :: Rule
+header_body =
+  "header_body"
+    =:= arr
+      [ "block_number" ==> VUInt
+      , "slot" ==> VUInt
+      , "prev_hash" ==> hash32
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: `prev_hash` is optional
```suggestion
      , "prev_hash" ==> hash32 // VNil
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 11]
```diff
@@ -0,0 +1,971 @@
...
+      , "issuer_vkey" ==> vkey
+      , "vrf_vkey" ==> vkey
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]:
```suggestion
      , "vrf_vkey" ==> vrf_vkey
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 12]
```diff
@@ -0,0 +1,971 @@
...
+  "operational_cert"
+    =:= arr
+      [ "hot_vkey" ==> kes_vkey
+      , "sequence_number" ==> VUInt
+      , "kes_period" ==> VUInt
+      , "sigma" ==> signature
+      ]
+
+protocol_version :: Named Group
+protocol_version = "protocol_version" =:~ grp [a major_protocol_version, a VUInt]
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Can we define it like that, instead of using `(arr [a protocol_version])` at use sites?
```suggestion
protocol_version = "protocol_version" =:~ arr [a (grp [a major_protocol_version, a VUInt])]
```

This way we would be able to use it as before [`a protocol_version`](https://github.com/IntersectMBO/cardano-ledger/blob/f4119c9cd27e69943dcf91d22d0f9905be32fa75/eras/conway/impl/cddl-files/conway.cddl#L37)

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 13]
```diff
@@ -0,0 +1,971 @@
...
+
+next_major_protocol_version :: Rule
+next_major_protocol_version = "next_major_protocol_version" =:= (10 :: Integer)
+
+major_protocol_version :: Rule
+major_protocol_version = "major_protocol_version" =:= (1 :: Integer)
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: This should be a list from [`1..next_major_protocol_version`](https://github.com/IntersectMBO/cardano-ledger/blob/f4119c9cd27e69943dcf91d22d0f9905be32fa75/eras/conway/impl/cddl-files/conway.cddl#L49)
```suggestion
major_protocol_version = "major_protocol_version" =:= (1 :: Integer)
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs [hunk 14]
```diff
@@ -0,0 +1,971 @@
...
+
+transaction_body :: Rule
+transaction_body =
+  "transaction_body"
+    =:= mp
+      [ idx 0 ==> set transaction_input
+      , idx 1 ==> arr [0 <+ a transaction_output]
+      , idx 2 ==> coin
+      , opt (idx 3 ==> VUInt)
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: We now have defined [`slot_no`](https://github.com/IntersectMBO/cardano-ledger/blob/f4119c9cd27e69943dcf91d22d0f9905be32fa75/eras/conway/impl/cddl-files/conway.cddl#L589)
```suggestion
      , opt (idx 3 ==> slot_no)
```

**Review by lehins — CHANGES_REQUESTED** [review]
This looks great!!! I am loving it!!! :heart: 

Looks like you've used an older CDDL spec as reference, could you please go through and compare to an up-to-date version and make necessary adjustments?

I haven't gone through the whole PR, but as soon as all current comments are addressed I'll do a thorough review.

This PR pretty much takes care of #4125 and #4126

There are more to go: #4461 and #4462 and then some other ideas that we've discussed

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs (hunk 1 above)
Comment (nc6) [reply eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs]: Whoops. Yeah, that was a hangover from testing when I was commenting things out. Will fix

File: eras/conway/impl/cardano-ledger-conway.cabal (hunk 7 above)
Comment (nc6) [reply eras/conway/impl/cardano-ledger-conway.cabal]: It already is in testlib. Should one update the version when the testlib library is updated but nothing else?

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs (hunk 6 above)
Comment (nc6) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Yes, the `sized` function works exactly in this way. Updated to reflect this

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs (hunk 12 above)
Comment (nc6) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Yes, this is fine. My translation of the Conway CDDL preceded 280ace3ed7, hence this change.

File: eras/conway/impl/cardano-ledger-conway.cabal (hunk 7 above)
Comment (nc6) [reply eras/conway/impl/cardano-ledger-conway.cabal]: Oh, sorry, you mean that Conway depends on Cuddle. I can move to testlib, I had placed it in the library merely to reflect the current location of the CDDL

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs (hunk 12 above)
Comment (nc6) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Though there's not really any point in the extra group indirection here.

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs (hunk 13 above)
Comment (nc6) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Ah, currently huddle doesn't support this. No reason it can't, but for now I've inlined it and added a TODO.

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs (hunk 4 above)
Comment (nc6) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Yes, it never helped anyway

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs (hunk 5 above)
Comment (nc6) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Yes, this should all eventually be modularised - that's the next thing on my TODO list!

Comment (nc6) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: And we can already generate examples without using the cddl-tool!

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs (hunk 2 above)
Comment (nc6) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Isn't `distinct` a hack, though? Precisely the hack that we're referring to.

Comment (teodanciu) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Sorry, I assumed we don't need the hack anymore, but indeed, since we still use distinct in the type itself, the comment makes sense.

File: eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs (hunk 5 above)
Comment (teodanciu) [reply eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs]: Aha, I meant we can remove the tool once we have the definitions for all eras. Nice!

File: eras/conway/impl/cardano-ledger-conway.cabal (hunk 7 above)
Comment (lehins) [reply eras/conway/impl/cardano-ledger-conway.cabal]: I'll be a bit more precise. This is where it needs to go: https://github.com/IntersectMBO/cardano-ledger/blob/8cff9cfcf23686bbde8bfa5dc0682630af38c517/eras/conway/impl/cardano-ledger-conway.cabal#L112

Comment (lehins) [reply eras/conway/impl/cardano-ledger-conway.cabal]: Current location of CDDL is in [the package](https://github.com/IntersectMBO/cardano-ledger/blob/8cff9cfcf23686bbde8bfa5dc0682630af38c517/eras/conway/impl/cardano-ledger-conway.cabal#L16), not the library. So, sublibraries have full access to those files as well

**Comment by nc6** [comment]
Note that I have #4461 addressed, but planned to include that in a subsequent PR to avoid inflating this one.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs [hunk 15]
```diff
@@ -0,0 +1,1008 @@
...
+block :: Rule
+block =
+  "block"
+    =:= arr
+      [ a header
+      , "transaction_bodies" ==> arr [0 <+ a transaction_body]
+      , "transaction_witness_sets"
+          ==> arr [0 <+ a transaction_witness_set]
+      , "auxiliary_data_set"
+          ==> mp [0 <+ asKey transaction_index ==> auxiliary_data]
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:35]: This is interesting. In Huddle `==>` operator is used for two different types of syntax:
```
[ key : value ]
```
and
```
{ key => value }
```
Do they have a similar meaning in CDDL or was it just possible to do this for convenience in Huddle?

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs [hunk 16]
```diff
@@ -0,0 +1,1008 @@
...
+      , a transaction_witness_set
+      , a VBool
+      , a (auxiliary_data // VNil)
+      ]
+
+transaction_index :: Rule
+transaction_index = "transaction_index" =:= VUInt `sized` (2 :: Word64)
+
+header :: Rule
+header = "header" =:= arr [a header_body, "body_signature" ==> kes_signature]
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:53]: It is a bit easier to see that those are separate fields this way
```suggestion
header = "header" =:= arr 
  [ a header_body
  , "body_signature" ==> kes_signature
  ]
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs [hunk 17]
```diff
@@ -0,0 +1,1008 @@
...
+
+policy_hash :: Rule
+policy_hash = "policy_hash" =:= scripthash
+
+parameter_change_action :: Named Group
+parameter_change_action =
+  "parameter_change_action"
+    =:~ grp
+      [ 0
+      , gov_action_id // VNil
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:166]: CDDL is too confusing! So, in the current spec it is defined as `gov_action_id / null`, where it is one `/`, while in Huddle it uses `//`

But for the definition of the actual gov action it is defined as:
```cddl
gov_action =
  [ parameter_change_action
  // hard_fork_initiation_action
  // treasury_withdrawals_action
   ....
```
with usage of `//`, while for Huddle it is the opposite and `/` is used.

Moreover, to make things even more confusing `voting_procedure` uses a single `/` in both the original CDDL spec and in Huddle.


Could you please explain what is going on with those `/` vs `//` in Huddle and CDDL? My confusion probably come from my limited understanding of these operators in CDDL itself, but this inconsistency and lack of understanding is bothering me.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs [hunk 18]
```diff
@@ -0,0 +1,1008 @@
...
+          / vote_reg_deleg_cert
+          / stake_vote_reg_deleg_cert
+          / auth_committee_hot_cert
+          / resign_committee_cold_cert
+          / reg_drep_cert
+          / unreg_drep_cert
+          / update_drep_cert
+      ]
+
+stake_registration :: Named Group
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:301]: Is there support in Huddle for comments?

It would be really useful to mark `stake_registration` and `stake_deregistration` as to  be deprecated in the future era.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs [hunk 19]
```diff
@@ -0,0 +1,1008 @@
...
+  "transaction"
+    =:= arr
+      [ a transaction_body
+      , a transaction_witness_set
+      , a VBool
+      , a (auxiliary_data // VNil)
+      ]
+
+transaction_index :: Rule
+transaction_index = "transaction_index" =:= VUInt `sized` (2 :: Word64)
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:50]: It would probably be useful to add `IsSize` instance for `Word` as well in Cuddle. But that is a minor suggestion for the future, rather that something we need to do right now
```suggestion
transaction_index = "transaction_index" =:= VUInt `sized` (2 :: Word)
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs [hunk 20]
```diff
@@ -0,0 +1,1008 @@
...
+      , opt $ idx 4 ==> nonempty_set plutus_data
+      , opt $ idx 5 ==> redeemers
+      , opt $ idx 6 ==> nonempty_set plutus_v2_script
+      , opt $ idx 7 ==> nonempty_set plutus_v3_script
+      ]
+
+-- The real type of  plutus_v1_script, plutus_v2_script and plutus_v3_script is bytes.
+-- However, because we enforce uniqueness when many scripts are supplied,
+-- we need to hack around for tests in order to avoid generating duplicates,
+-- since the cddl tool we use for roundtrip testing doesn't generate distinct collections.
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:551]: I think it is enough to have the same comment in Huddle.
```suggestion
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs [hunk 21]
```diff
@@ -0,0 +1,1008 @@
...
+    // tag 125 (arr [0 <+ a x])
+    // tag 126 (arr [0 <+ a x])
+    // tag 127 (arr [0 <+ a x])
+    -- similarly for tag range: 6.1280 .. 6.1400 inclusive
+    // tag 102 (arr [a VUInt, a $ arr [0 <+ a x]])
+
+redeemers :: Rule
+redeemers =
+  "redeemers"
+    =:= sarr
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:602]: Same here: adding a Huddle comment about deprecation would be very valuable

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs [hunk 22]
```diff
@@ -0,0 +1,1008 @@
...
+script_all :: Named Group
+script_all = "script_all" =:~ grp [1, a (arr [0 <+ a native_script])]
+
+script_any :: Named Group
+script_any = "script_any" =:~ grp [2, a (arr [0 <+ a native_script])]
+
+script_n_of_k :: Named Group
+script_n_of_k =
+  "script_n_of_k"
+    =:~ grp [3, "n" ==> VUInt, a (arr [0 <+ a native_script])]
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:751]: This is actually incorrect in the original spec. It is a signed 64bit `Int`. Could you please fix it here and in the spec. It's fine if you'd prefer not to change that in this PR then just feel free to convert this comment into a ticket and we'll take care of it later
```suggestion
    =:~ grp [3, "n" ==> int64, a (arr [0 <+ a native_script])]
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs [hunk 23]
```diff
@@ -0,0 +1,1008 @@
...
+pool_metadata_hash = "pool_metadata_hash" =:= hash32
+
+-- To compute a script hash, note that you must prepend
+-- a tag to the bytes of the script before hashing.
+-- The tag is determined by the language.
+-- The tags in the Conway era are:
+--   "\x00" for multisig scripts
+--   "\x01" for Plutus V1 scripts
+--   "\x02" for Plutus V2 scripts
+--   "\x03" for Plutus V3 scripts
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:855]: Ditto
```suggestion
```

**Review by lehins — APPROVED** [review]
Few minor suggestions and question, but nothing critical.

Cuddle/Huddle is an amazing piece of work! Thank you for getting us to this point @nc6!

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs (hunk 15 above)
Comment (nc6) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:35]: In CDDL the `key : value` syntax is short for `"key" ==> value`. Strings have to be quoted in Haskell, so they're merged.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs (hunk 17 above)
Comment (nc6) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:166]: So, in CDDL, `/` is a choice between types (concretely, between `Type1` values, to make a `Type0`). `//` allows choice between groups. We can illustrate the difference:

```
foo = [ 0 / 1, uint // 2 /3, tstr ]
```
This construction would match either of the following:
```
[0, 3]
[2, "Hello World"]
```

 In other words, the `//` binds less strongly than `,` in CDDL.

In Haskell, of course, we cannot have syntax inside an array which binds stronger than the `,`. so we have to do things a little differently. The way this is handled at the moment is that `//` has special treatment for arrays/groups, where it will, instead of creating a type-level choice, merge the two arrays/groups/maps into a single one containing a group choice.

So, two questions remain: why is this operator `//` instead of `/` (whose semantics it more closely matches), and what is the `/` operator in Huddle?

The former is largely accidental: the group choice operator got expanded to other cases.

What about the latter? Well, that was mostly intended to handle the case of array entries. But looking at it again, I think it's a mistake. The case it was used for can be replaced by updating fixity declarations.

So I think I will update this with the following intention: all choices within huddle will be written with `/`, and I will incorporate most of this comment into the documentation to clarify it.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs (hunk 18 above)
Comment (nc6) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:301]: Done

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs (hunk 19 above)
Comment (nc6) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:50]: https://github.com/input-output-hk/cuddle/issues/33
